beluga_ros/include/beluga_ros/point_cloud.hpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ROS_POINT_CLOUD_HPP
+#define BELUGA_ROS_POINT_CLOUD_HPP
+
+#include <range/v3/view/iota.hpp>
+
+#include <beluga/sensor/data/point_cloud.hpp>
+#include <beluga/views/take_evenly.hpp>
+#include <beluga_ros/messages.hpp>
+
+#include <sophus/se3.hpp>
+
+#include <Eigen/Dense>
+
+/**
+ * \file
+ * \brief Implementation of `sensor_msgs/PointCloud2` wrapper type.
+ */
+
+namespace beluga_ros {
+
+template <uint8_t T>
+struct DataType;
+
+template <>
+struct DataType<beluga_ros::msg::PointFieldF32> {
+  using iteratorType = float;
+  using eigenType = const Eigen::Matrix3Xf;
+};
+
+template <>
+struct DataType<beluga_ros::msg::PointFieldF64> {
+  using iteratorType = double;
+  using eigenType = const Eigen::Matrix3Xd;
+};
+
+/// Thin wrapper type for 3D `sensor_msgs/PointCloud2` messages.
+/// Assumes an XYZ... type message.
+/// Each field must have the same datatype.
+/// The point cloud can't have invalid values, i.e., it must be dense.
+template <uint8_t T = beluga_ros::msg::PointFieldF64>
+class PointCloud3 : public beluga::BasePointCloud<PointCloud3<T>> {
+ public:
+  /// PointCloud data fields type
+  using iteratorType = typename DataType<T>::iteratorType;
+  using eigenType = typename DataType<T>::eigenType;
+
+  /// Check type is float or double
+  static_assert(
+      std::is_same<iteratorType, float>::value || std::is_same<iteratorType, double>::value,
+      "Pointcloud3 only supports float or double datatype");
+
+  /// Constructor.
+  ///
+  /// \param cloud Point cloud message.
+  /// \param origin Point cloud frame origin in the filter frame.
+  explicit PointCloud3(beluga_ros::msg::PointCloud2ConstSharedPtr cloud, Sophus::SE3d origin = Sophus::SE3d())
+      : cloud_(std::move(cloud)), origin_(std::move(origin)) {
+    // Check there are not invalid values
+    if (!cloud_->is_dense)
```

Comment:
I think pointing out that the pointcloud was not dense if the user wanted to use a dense one is not important, since we are going to filter those values anyway. So I totally removed this verification  in d9a35872f8e3c100f73c56daefe5bbf4cd58716a